Source/Engine/Mandarin/Mandarin.cpp

```diff
@@ -23,991 +20,1236 @@
 }
 
-const BPMF BPMF::FromComposedString(const string& str)
-{
-    BPMF syllable;
-    vector<string> components = OVUTF8Helper::SplitStringByCodePoint(str);
-    for (vector<string>::iterator iter = components.begin() ; iter != components.end() ; ++iter) {
-	
-		const map<string, BPMF::Component>& charToComp = BopomofoCharacterMap::SharedInstance().characterToComponent;
-		map<string, BPMF::Component>::const_iterator result = charToComp.find(*iter);
-		if (result != charToComp.end())
-        	syllable += BPMF((*result).second);
-    }
-    
-    return syllable;
+BopomofoCharacterMap::BopomofoCharacterMap() {
+  characterToComponent[u8"ㄅ"] = BPMF::B;
+  characterToComponent[u8"ㄆ"] = BPMF::P;
+  characterToComponent[u8"ㄇ"] = BPMF::M;
+  characterToComponent[u8"ㄈ"] = BPMF::F;
+  characterToComponent[u8"ㄉ"] = BPMF::D;
+  characterToComponent[u8"ㄊ"] = BPMF::T;
+  characterToComponent[u8"ㄋ"] = BPMF::N;
+  characterToComponent[u8"ㄌ"] = BPMF::L;
+  characterToComponent[u8"ㄎ"] = BPMF::K;
+  characterToComponent[u8"ㄍ"] = BPMF::G;
+  characterToComponent[u8"ㄏ"] = BPMF::H;
+  characterToComponent[u8"ㄐ"] = BPMF::J;
+  characterToComponent[u8"ㄑ"] = BPMF::Q;
+  characterToComponent[u8"ㄒ"] = BPMF::X;
+  characterToComponent[u8"ㄓ"] = BPMF::ZH;
+  characterToComponent[u8"ㄔ"] = BPMF::CH;
+  characterToComponent[u8"ㄕ"] = BPMF::SH;
+  characterToComponent[u8"ㄖ"] = BPMF::R;
+  characterToComponent[u8"ㄗ"] = BPMF::Z;
+  characterToComponent[u8"ㄘ"] = BPMF::C;
+  characterToComponent[u8"ㄙ"] = BPMF::S;
+  characterToComponent[u8"ㄧ"] = BPMF::I;
+  characterToComponent[u8"ㄨ"] = BPMF::U;
+  characterToComponent[u8"ㄩ"] = BPMF::UE;
+  characterToComponent[u8"ㄚ"] = BPMF::A;
+  characterToComponent[u8"ㄛ"] = BPMF::O;
+  characterToComponent[u8"ㄜ"] = BPMF::ER;
+  characterToComponent[u8"ㄝ"] = BPMF::E;
+  characterToComponent[u8"ㄞ"] = BPMF::AI;
+  characterToComponent[u8"ㄟ"] = BPMF::EI;
+  characterToComponent[u8"ㄠ"] = BPMF::AO;
+  characterToComponent[u8"ㄡ"] = BPMF::OU;
+  characterToComponent[u8"ㄢ"] = BPMF::AN;
+  characterToComponent[u8"ㄣ"] = BPMF::EN;
+  characterToComponent[u8"ㄤ"] = BPMF::ANG;
+  characterToComponent[u8"ㄥ"] = BPMF::ENG;
+  characterToComponent[u8"ㄦ"] = BPMF::ERR;
+  characterToComponent[u8"ˊ"] = BPMF::Tone2;
+  characterToComponent[u8"ˇ"] = BPMF::Tone3;
+  characterToComponent[u8"ˋ"] = BPMF::Tone4;
+  characterToComponent[u8"˙"] = BPMF::Tone5;
+
+  for (std::map<std::string, BPMF::Component>::iterator iter =
+           characterToComponent.begin();
+       iter != characterToComponent.end(); ++iter)
+    componentToCharacter[(*iter).second] = (*iter).first;
 }
 
-const string BPMF::composedString() const
-{
-    string result;
-    #define APPEND(c) if (m_syllable & c) result += (*BopomofoCharacterMap::SharedInstance().componentToCharacter.find(m_syllable & c)).second
-    APPEND(ConsonantMask);
-    APPEND(MiddleVowelMask);
-    APPEND(VowelMask);
-    APPEND(ToneMarkerMask);
-    #undef APPEND
-    return result;
+#define ASSIGNKEY1(m, vec, k, val) \
```

Comment:
That really has no benefits, and can really be written now with things like `m[k] = std::vector{val1, val2, ...}`. Even back in those days it could have been written with `m[k].push_back(...); m[k].push_back(...);`. It was really just an oversight.

`ASSIGNKEYx` could use a better macro name (or better, as an inline function), and it really means "assign the key at state _x_ to the Bopmofo syllable".